File: go/monoidal.go
package combinators

type Object interface{}
type Morphism interface {
    Apply(Object) Object
}

type MonoidalCategory struct{}

func (mc MonoidalCategory) Id(obj Object) Morphism {
    return IdMorphism{obj}
}

func (mc MonoidalCategory) Compose(f, g Morphism) Morphism {
    return ComposedMorphism{f, g}
}

func (mc MonoidalCategory) Tensor(a, b Object) Object {
    return TensorObject{a, b}
}

func (mc MonoidalCategory) TensorMor(f, g Morphism) Morphism {
    return TensorMorphism{f, g}
}

type IdMorphism struct {
    Object Object
}

func (m IdMorphism) Apply(obj Object) Object {
    return m.Object
}

type ComposedMorphism struct {
    F, G Morphism
}

func (m ComposedMorphism) Apply(obj Object) Object {
    return m.F.Apply(m.G.Apply(obj))
}

type TensorObject struct {
    A, B Object
}

type TensorMorphism struct {
    F, G Morphism
}

func (m TensorMorphism) Apply(obj Object) Object {
    pair := obj.(TensorObject)
    return TensorObject{m.F.Apply(pair.A), m.G.Apply(pair.B)}
}

File: ocaml/lib/monoidal.ml
module type MonoidalCategory = sig
  type 'a obj
  type 'a mor
  val id : 'a obj -> 'a mor
  val compose : 'b mor -> 'a mor -> 'a obj -> 'b obj -> 'c mor
  val tensor : 'a obj -> 'b obj -> 'c obj
  val tensorMor : 'a mor -> 'b mor -> 'c mor
end

module MonoidalCategory : MonoidalCategory = struct
  type 'a obj = 'a
  type 'a mor = 'a -> 'a
  let id x = x
  let compose f g a b = f (g a)
  let tensor a b = (a, b)
  let tensorMor f g (a, b) = (f a, g b)
end